src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -217,18 +231,55 @@ private int LookupAggregatorStore(string[] tagKeys, object[] tagValues, int leng
             return aggregatorIndex;
         }
 
+        private bool TryGetUnusedMetricPoint(out int index)
+        {
+            if (this.metricPointIndex < MaxMetricPoints)
+            {
+                index = Interlocked.Increment(ref this.metricPointIndex);
+                if (index < MaxMetricPoints)
+                {
+                    return true;
+                }
+            }
+
+            if (this.temporality == AggregationTemporality.Delta)
+            {
+                for (int i = 1; i < MaxMetricPoints; i++)
+                {
+                    ref var metricPoint = ref this.metricPoints[i];
+                    if (metricPoint.MetricPointStatus == MetricPointStatus.CandidateForRemoval)
```

Comment:
I think the approach of freeing up space on Update path will lead to incorrect behavior.
say K1 is the key to the dictionary, pointing to MP P1 which is candidateforremoval.

Following two threads:
Thread 1: Update(K1,100) 
Looks up dictionary and finds K1->P1 mapping, so decides to Update P1.
Thread: 2: Update(K2, 135) - K2 is not found. out of MPs. So try to free up, and it can decide to free up P1. It frees up P1, and removes (K1->P1) and put K2->P1. Updates P1 with 135, status to PendingCollect.
Thread1 - is unaware that P1 is taken up by someone else, and falsely stores its update to P1.

Reply:
Ugh, yep agreed this is a problem, I've taken the route you suggested of validating in the update path that the MP keys/values are the ones expected. It might not be the prettiest solutions 🥴. I believe the fix will prevent Thread 1 from updating the wrong MP, but there may still be the possibility that the update on Thread 1 gets lost, but I think this is not nearly as bad.